src/foundry/foundry.js/roll.d.ts

```diff
@@ -1,501 +1,515 @@
-// TODO: Remove when updating this class!!!
-// eslint-disable-next-line
-// @ts-nocheck
-
-/**
- * This class provides an interface and API for conducting dice rolls.
- * The basic structure for a dice roll is a string formula and an object of data against which to parse it.
- *
- * @typeParam D - the type of data object against which to parse attributes within the formula
- *
- * @example
- * ```typescript
- * // Attack with advantage!
- * let r = new Roll("2d20kh + @prof + @strMod", {prof: 2, strMod: 4});
- *
- * // The parsed terms of the roll formula
- * console.log(r.terms);    // [Die, OperatorTerm, NumericTerm, OperatorTerm, NumericTerm]
- *
- * // Execute the roll
- * r.evaluate();
- *
- * // The resulting equation after it was rolled
- * console.log(r.result);   // 16 + 2 + 4
- *
- * // The total resulting from the roll
- * console.log(r.total);    // 22
- * ```
- */
-declare class Roll<D extends object = {}> {
-  /**
-   * @param formula - The string formula to parse
-   * @param data    - The data object against which to parse attributes within the formula
-   *                  (default: `{}`)
-   */
-  constructor(formula: string, data?: D, options?: Roll['options']);
-
-  /**
-   * The original provided data
-   */
-  data: D;
+import { ConfiguredDocumentClass } from '../../types/helperTypes';
 
+declare global {
   /**
-   * Options which modify or describe the Roll
+   * This class provides an interface and API for conducting dice rolls.
+   * The basic structure for a dice roll is a string formula and an object of data against which to parse it.
+   *
+   * @typeParam D - the type of data object against which to parse attributes within the formula
+   *
+   * @example
+   * ```typescript
+   * // Attack with advantage!
+   * let r = new Roll("2d20kh + @prof + @strMod", {prof: 2, strMod: 4});
+   *
+   * // The parsed terms of the roll formula
+   * console.log(r.terms);    // [Die, OperatorTerm, NumericTerm, OperatorTerm, NumericTerm]
+   *
+   * // Execute the roll
+   * r.evaluate();
+   *
+   * // The resulting equation after it was rolled
+   * console.log(r.result);   // 16 + 2 + 4
+   *
+   * // The total resulting from the roll
+   * console.log(r.total);    // 22
+   * ```
    */
-  options: Partial<Roll.Options>;
+  class Roll<D extends object = {}> {
+    /**
+     * @param formula - The string formula to parse
+     * @param data    - The data object against which to parse attributes within the formula
+     *                  (default: `{}`)
+     */
+    constructor(formula: string, data?: D, options?: Roll['options']);
 
-  /**
-   * The identified terms of the Roll
-   */
-  terms: RollTerm[];
+    /**
+     * The original provided data
+     */
+    data: D;
```

Comment:
Technically, this would be of type `ReturnType<this['_prepareData']>`, which may be different from the `data` passed to the constructor in custom `Roll` classes.

Reply:
`TS4105: Private or protected member '_prepareData' cannot be accessed on a type parameter.`

`ReturnType<Roll<D>['_prepareData']>` would be possible, but pointless.